### PR TITLE
Open Source Delta Maintainability Model Support

### PR DESCRIPTION
**Pull request by avandeursen**
This pull request extends PyDriller with support for the Delta Maintainability Model. It offers metrics to characterize the change in maintainability caused by commit, focusing on the proportion of *low-risk change* in a commit. 

The pull request offers:

- `docs/deltamaintainability.rst`: Documentation of the origins of the model, its usage in PyDriller, and the `Commit` API extensions with `dmm_unit_size`, `dmm_unit_complexity`, and `dmm_unit_interfacing`.
- `pydriller/domain/commit.py`: Implementation of the model in the `Commit` and `Modification` classes
- `tests/test_dmm.py`: Series of test cases covering all corner cases and all branches
- `test-repos.zip`: Extended with test repo  available at https://github.com/avandeursen/dmm-test-repo

See also Marco di Biase, Ayushi Rastogi, Magiel Bruntink, and Arie van Deursen. The Delta Maintainability Model: measuring maintainability of fine-grained code changes. IEEE/ACM International Conference on Technical Debt (TechDebt) at ICSE 2019, pp 113-122 ([preprint and doi](https://pure.tudelft.nl/portal/en/publications/the-delta-maintainability-model-measuring-maintainability-of-finegrained-code-changes(6ff67dee-2781-47d7-916f-bd36c5b61beb).html)).

### CODE REVIEW

File: pydriller/domain/commit.py [hunk 1]
```diff
@@ -562,6 +590,161 @@ def _get_branches(self):
...
+    Threshold used in the Delta Maintainability Model to establish whether a method
+    is low risk in terms of its size.
+    """
+    @staticmethod
+    def _low_risk_unit_size(method) -> bool:
+        """
+        Predicate indicating whether a method is low risk in terms of its size,
+        i.e., whether its number of lines of code remains on or below the low risk threshold.
+        """
+        return method.nloc <= Commit._UNIT_SIZE_LOW_RISK_THRESHOLD
```
Comment (ishepard) [inline pydriller/domain/commit.py:605]: What if we bring this to the Method class? And we make it public. As well as `UNIT_SIZE_LOW_RISK_THRESHOLD`

File: pydriller/domain/commit.py [hunk 2]
```diff
@@ -562,6 +590,161 @@ def _get_branches(self):
...
+    Threshold used in the Delta Maintainability Model to establish whether a method
+    is low risk in terms of its cyclomatic complexity.
+    """
+    @staticmethod
+    def _low_risk_unit_complexity(method: Method) -> bool:
+        """
+        Predicate indicating whether a method is low risk in terms of cyclomatic complexity,
+        i.e., whether the cyclomatic complexity remains on or below the low risk threshold.
+        """
+        return method.complexity <= Commit._UNIT_COMPLEXITY_LOW_RISK_THRESHOLD
```
Comment (ishepard) [inline pydriller/domain/commit.py:618]: What if we bring this to the Method class? And we make it public. As well as UNIT_COMPLEXITY_LOW_RISK_THRESHOLD

File: pydriller/domain/commit.py [hunk 3]
```diff
@@ -562,6 +590,161 @@ def _get_branches(self):
...
+    is low risk in terms of its interface.
+    """
+    @staticmethod
+    def _low_risk_unit_interfacing(method: Method) -> bool:
+        """
+        Predicate indicating whether a method is low risk in terms of interfacing,
+        i.e., whether the number of parameters of this method remains on or below
+        the low risk threshold.
+        """
+        return len(method.parameters) <= Commit._UNIT_INTERFACING_LOW_RISK_THRESHOLD
```
Comment (ishepard) [inline pydriller/domain/commit.py:632]: What if we bring this to the Method class? And we make it public. As well as UNIT_INTERFACING_LOW_RISK_THRESHOLD

File: pydriller/domain/commit.py (hunk 1 above)
Comment (ishepard) [reply pydriller/domain/commit.py:605]: So a user can see if a method is "low_risk_unit_size" or "high_risk_unit_size"

File: pydriller/domain/commit.py [hunk 4]
```diff
@@ -562,6 +590,161 @@ def _get_branches(self):
             branches.add(branch.strip().replace('* ', ''))
         return branches
 
+
+    _UNIT_SIZE_LOW_RISK_THRESHOLD = 15
```
Comment (ishepard) [inline pydriller/domain/commit.py:594]: Does this threshold come from the paper and/or SIG MM? 
Maybe we should add a link or something. I am expecting someone to ask "why 15?" 😄 
Similarly, check the other metrics.

**Comment by avandeursen** [comment]
OK, thanks, sounds good.

The thresholds are now explained in the documentation and come from SIG/MM as described in the paper. I'll copy part of that info to the comments.

Moving the `low_risk...`  predicates as methods of the `Method` class and making them public makes sense. Will do.

Passing a flag instead of a predicate would amount to creating an enumeration (UNIT_SIZE, UNIT_COMPLEXITY, UNIT_INTERFACING), and then mapping that flag to the corresponding Callable in the Method class. Not sure this indeed leads to better code.

**Comment by ishepard** [comment]
> The thresholds are now explained in the documentation and come from SIG/MM as described in the paper. I'll copy part of that info to the comments.
> Moving the `low_risk...` predicates as methods of the `Method` class and making them public makes sense. Will do.

Awesome, thanks!

> Passing a flag instead of a predicate would amount to creating an enumeration (UNIT_SIZE, UNIT_COMPLEXITY, UNIT_INTERFACING), and then mapping that flag to the corresponding Callable in the Method class. Not sure this indeed leads to better code, but will explore.

Indeed, I am not sure..I will leave it to you! Thanks!

**Comment by ishepard** [comment]
LGTM! 
Thanks @avandeursen!
I will release a new version of Pydriller soon.
